syntax = "proto3";

package wfa.any_sketch.crypto;

import "wfa/measurement/api/v1alpha/sketch.proto";

option java_package = "org.wfanet.anysketch.crypto";
option java_multiple_files = true;

// Public keys of the ElGamal cipher.
message ElGamalPublicKeys {
  // Public key g (generator) of the ElGamal cipher.
  bytes el_gamal_g = 1;
  // Public key y (g^sk) of the ElGamal cipher.
  bytes el_gamal_y = 2;
}

// The request to encrypt a sketch.
message EncryptSketchRequest {
  // The input sketch
  wfa.measurement.api.v1alpha.Sketch sketch = 1;
  // Public keys of the ElGamal cipher used to encrypt the sketch
  ElGamalPublicKeys el_gamal_keys = 2;
  // The elliptical curve to work on.
  int64 curve_id = 3;
  // Maximum value to encrypt. Every value greater than the maximum_value
  // is mapping to the encryption of the maximum_value.
  int32 maximum_value = 4;
}

// Response of the EncryptSketch method.
message EncryptSketchResponse {
  // The encrypted sketch
  bytes encrypted_sketch = 1;
}